### PR TITLE
Bulk export

### PR DESCRIPTION
**Pull request by spiderxm**
Adding feature for user to export data as json
Screenshot 2021-06-15 at 4 51 41 PM

### CODE REVIEW

**Comment by spiderxm** [comment]
@Ali-Razmjoo @dhirensr Please review

File: web/static/js/explore.js [hunk 1]
```diff
@@ -275,18 +275,35 @@ function get_event_data(api_endpoint, column_list, api_params) {
...
+                },
+                className: "btn btn-info btn-sm"
+            },
+
+            {
+                text: 'JSON',
+                action: function (e, dt, button, config) {
+                    const filename = get_export_fileName('json') + ".json"
+                    $.fn.dataTable.fileSave(
+                        new Blob([JSON.stringify(table.rows().data().toArray(), undefined, 2)], {
```
Comment (dhirensr) [inline web/static/js/explore.js:299]: hello, there's a data consistency isssue.
I see a lot of entries in the explorer UI, but when I export in json I only get 10 records from the table's first page.

**Comment by dhirensr** [comment]
@spiderxm : I checked it now on latest version of this branch,
2 problems I'm facing
1. files are not getting saved as ".csv" and instead "Honeypot-data-csv-1624359617127csv"
2.  so the csv and excel for example shows all the data now, but the json still looks smaller with less data.

can you also test that?

File: web/static/js/explore.js [hunk 2]
```diff
@@ -252,6 +252,74 @@ $.fn.dataTable.Api.register( 'clearPipeline()', function () {
...
+                case "JSON": {
+                    const filename = get_export_fileName('json') + ".json"
+                    const dataStr = "data:text/json;charset=utf-8," + encodeURIComponent(JSON.stringify(data, undefined, 2));
+                    const downloadAnchorNode = document.createElement('a');
+                    downloadAnchorNode.setAttribute("href", dataStr);
+                    downloadAnchorNode.setAttribute("download", filename);
+                    document.body.appendChild(downloadAnchorNode);
+                    downloadAnchorNode.click();
+                    downloadAnchorNode.remove();
+                    break
```
Comment (dhirensr) [inline web/static/js/explore.js:284]: break; ? better not to rely on Automatic Semicolon Insertion

File: web/static/js/explore.js [hunk 3]
```diff
@@ -252,6 +252,74 @@ $.fn.dataTable.Api.register( 'clearPipeline()', function () {
...
+                        csvFileData += `"${column_list[column].data}"` + ","
+                    }
+                    csvFileData += "\n"
+                    for (let index = 0; index < data.length; index++) {
+                        for (let column in column_list) {
+                            csvFileData += `"${data[index][column_list[column].data]}"` + ","
+                        }
+                        csvFileData += "\n"
+                    }
+                    const filename = get_export_fileName('csv') + "csv"
```
Comment (dhirensr) [inline web/static/js/explore.js:298]: const filename = get_export_fileName('csv') + ".csv"

**Comment by spiderxm** [comment]
@dhirensr Will fix that.

File: web/static/js/explore.js [hunk 4]
```diff
@@ -252,6 +252,74 @@ $.fn.dataTable.Api.register( 'clearPipeline()', function () {
...
+
+                    const dataStr = "data:text/csv;charset=utf-8," + encodeURIComponent(csvFileData);
+                    const downloadAnchorNode = document.createElement('a');
+                    downloadAnchorNode.setAttribute("href", dataStr);
+                    downloadAnchorNode.setAttribute("download", filename);
+                    document.body.appendChild(downloadAnchorNode);
+                    downloadAnchorNode.click();
+                    downloadAnchorNode.remove();
```
Comment (dhirensr) [inline web/static/js/explore.js:306]: is this needed? i checked few examples where remove wasn't used
if needed then let it be, am just asking out of curiosity

Comment (spiderxm) [reply web/static/js/explore.js:306]: ill check

**Comment by dhirensr** [comment]
@spiderxm : when I'm filtering data, i get less results in json, whereas csv and excel have consistent results. is it also for you?

**Comment by spiderxm** [comment]
> @spiderxm : when I'm filtering data, i get less results in json, whereas csv and excel have consistent results. is it also for you?

I didn't face issue like that. Data is consistent.

**Comment by spiderxm** [comment]
With ElasticSearch we can only request 10,000 records at max.
So if count is greater than that. We would get only 10,000 records

**Review by dhirensr — APPROVED** [review]
when I'm filtering data, i get less results in json, whereas csv and excel have consistent results. 
@Ali-Razmjoo : you can test that and then approve please

**Comment by Ali-Razmjoo** [comment]
It is a great feature, but it exports the data with false positives from the table. Can you export data from API calls directly instead of from the table? 

Thanks.

File: web/static/js/explore.js [hunk 5]
```diff
@@ -252,6 +252,72 @@ $.fn.dataTable.Api.register( 'clearPipeline()', function () {
...
+
+/**
+ * Function fetches log data based on params specified by user and download data in specified fileType
+ * @param {*} api_endpoint : API endpoint URL
+ * @param {*} column_list : List of Columns for the selected event type
+ * @param {*} api_params  : GET parameters for the API call
+ * @param {*} fileType    : File Type ( Json, Excel, Csv)
+ */
+function downloadLogData(api_endpoint, column_list, api_params, fileType) {
+    $.ajax({
```
Comment (spiderxm) [inline web/static/js/explore.js:264]: Data is fetched directly from the api rather than the table.

Comment (spiderxm) [reply web/static/js/explore.js:264]: @Ali-Razmjoo. The buttons are added inside the table. But on choosing on one. Data is fetched from api based on user filter and then converted to appropriate format and downloaded.

**Review by Ali-Razmjoo — APPROVED** [review]

**Comment by Ali-Razmjoo** [comment]
great job, thanks.

**Comment by Ali-Razmjoo** [comment]
I've changed the PR to elasticsearch branch...  some tests needed new codes that exist theere.

**Pull request merged**